crates/wit-component/src/validation.rs

```diff
@@ -154,22 +165,144 @@ pub fn validate_module<'a>(
+        let payload = payload?;
+        match validator.payload(&payload)? {
+            ValidPayload::End(tys) => {
+                types = Some(tys);
+                break;
+            }
+            ValidPayload::Func(validator, body) => {
+                funcs.push((validator, body));
+            }
+            _ => {}
+        }
+
+        match payload {
+            Payload::Version { encoding, .. } if encoding != Encoding::Module => {
+                bail!("data is not a WebAssembly module");
+            }
+
+            Payload::ImportSection(s) => {
+                for import in s {
+                    let import = import?;
+                    match import.ty {
+                        TypeRef::Func(ty) => {
+                            let map = match import_funcs.entry(import.module) {
+                                Entry::Occupied(e) => e.into_mut(),
+                                Entry::Vacant(e) => e.insert(IndexMap::new()),
+                            };
+
+                            assert!(map.insert(import.name, ty).is_none());
+                        }
+
+                        // A memory is allowed to be imported into the adapter
+                        // module so that's skipped here
+                        TypeRef::Memory(_) => {
+                            ret.needs_memory =
+                                Some((import.module.to_string(), import.name.to_string()));
+                        }
+
+                        _ => {
+                            bail!("adapter module is only allowed to import functions and memories")
+                        }
+                    }
+                }
+            }
+            Payload::ExportSection(s) => {
+                for export in s {
+                    let export = export?;
+
+                    match export.kind {
+                        ExternalKind::Func => {
+                            export_funcs.insert(export.name, export.index);
+                            if export.name == "cabi_realloc" {
+                                ret.has_realloc = true;
+                            }
+                        }
+                        _ => continue,
+                    }
+                }
+            }
+            _ => continue,
+        }
+    }
+
+    let mut resources = Default::default();
+    for (validator, body) in funcs {
+        let mut validator = validator.into_validator(resources);
+        validator.validate(&body)?;
+        resources = validator.into_allocations();
+    }
+
+    let types = types.unwrap();
+    for (name, funcs) in &import_funcs {
+        if *name != interface.name {
+            bail!(
+                "adapter module imports from `{name}` which does not match \
+                 its interface `{}`",
+                interface.name
+            );
+        }
+        ret.required_funcs = validate_imported_interface(interface, name, funcs, &types)?;
```

Comment:
I don't understand why this assignment is performed in each iteration of this function. I think for all our examples import_funcs only has one element, but I don't think this is the desired behavior if there are multiple elements there?

Reply:
Ah this is just a confusingly written loop, basically it requires at most one import since we only have one interface-per-adapter right now, and the loop validates that but in a non-obvious way, I'll update.